pkg/trace/stats/client_stats_aggregator.go

```diff
@@ -0,0 +1,283 @@
+package stats
+
+import (
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/trace/info"
+	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+	"github.com/DataDog/datadog-agent/pkg/trace/watchdog"
+)
+
+const (
+	aggBucketSize           = uint64(1e9)
+	bucketDuration          = uint64(1e10) // 10s
+	oldestBucketStartSecond = uint64(20)   // 20s
+	distributionsPayload    = "distributions"
+	aggregatedCountPayload  = "counts"
+)
+
+// ClientStatsAggregator aggregates client stats payloads on 1s buckets.
+// If a single payload is received on a bucket, this Aggregator is a passthrough.
+// If two ore more payloads collide, all payloads will be sent with hits and errors
+// set to 0. An aggregation will be started for Hits/Errors/Duration counts.
+// The Aggregation will be sent through a new stat payload with no Histograms.
+type ClientStatsAggregator struct {
+	In      chan pb.ClientStatsPayload
+	out     chan pb.StatsPayload
+	buckets map[uint64]*bucket // bucketts used to aggregate client stats
+
+	flushTicker   *time.Ticker
+	oldestTs      uint64
+	agentEnv      string
+	agentHostname string
+
+	exit chan struct{}
+	done chan struct{}
+}
+
+// NewClientStatsAggregator initializes a new aggregator ready to be started
+func NewClientStatsAggregator(conf *config.AgentConfig, out chan pb.StatsPayload) *ClientStatsAggregator {
+	oldestTSUnix := uint64(time.Now().Add(-10 * time.Second).Unix())
+	return &ClientStatsAggregator{
+		flushTicker:   time.NewTicker(time.Second),
+		In:            make(chan pb.ClientStatsPayload, 10),
+		buckets:       make(map[uint64]*bucket, 20),
+		out:           out,
+		agentEnv:      conf.DefaultEnv,
+		agentHostname: conf.Hostname,
+		oldestTs:      oldestTSUnix,
+	}
+}
+
+// Start starts the aggregator.
+func (a *ClientStatsAggregator) Start() {
+	go func() {
+		defer watchdog.LogOnPanic()
+		a.run()
+	}()
+}
+
+// Stop stops the aggregator.
+func (a *ClientStatsAggregator) Stop() {
+	close(a.exit)
+	<-a.done
+}
+
+func (a *ClientStatsAggregator) run() {
+	for {
+		select {
+		case t := <-a.flushTicker.C:
+			a.flushOnTime(t)
+		case input := <-a.In:
+			a.add(time.Now(), input)
+		case <-a.exit:
+			a.flushAll()
+			close(a.done)
+		}
+	}
+}
+
+func (a *ClientStatsAggregator) flushOnTime(t time.Time) {
```

Comment:
```suggestion
// flushOnTime flushes all buckets up to t, except the last one.
func (a *ClientStatsAggregator) flushOnTime(t time.Time) {
```
My documentation may be wrong, but please apply it corrected.